test/utils.ts

```diff
@@ -82,19 +101,35 @@ export function test<T extends ValidTestCase>(
 ): T extends { errors: InvalidTestCaseError[] | number }
   ? InvalidTestCase
   : ValidTestCase {
-  // @ts-expect-error -- simplify testing
-  return {
-    filename: TEST_FILENAME,
-    ...t,
-    languageOptions: {
-      ...t.languageOptions,
-      parserOptions: {
-        sourceType: 'module',
-        ecmaVersion: 9,
-        ...t.languageOptions?.parserOptions,
-      },
-    },
-  }
+  // @ts-expect-error simplify testing
+  return createRuleTestCase(t)
+}
+
+type GetRuleType<TRule> =
+  TRule extends RuleModule<infer MessageIds, infer Options>
+    ? {
+        messageIds: MessageIds
+        options: Options
+      }
+    : never
+
+export function createRuleTestCaseFunction<
+  TRule extends RuleModule<string, unknown[]>,
+  TData extends GetRuleType<TRule> = GetRuleType<TRule>,
+  TTestCase extends
+    | TSESLintValidTestCase<TData['options']>
+    | TSESLintInvalidTestCase<TData['messageIds'], TData['options']> =
+    | TSESLintValidTestCase<TData['options']>
+    | TSESLintInvalidTestCase<TData['messageIds'], TData['options']>,
+>(): <
```

Comment:
I don't like the idea of `createRuleTestCaseFunction<typeof rule>()`. I personally prefer things like `createRuleTestCaseFunction(_rule)` (I don't have to write an extra `typeof` here. Unit tests don't care about stripping type-only imports for tree-shaking anyway). What do you think?

And is it possible to deduce the `TRule` type based on the arg here?

Reply:
I personally use the type parameter syntax to explicit the fact that the parameter is only used at `typescript` level and does not affect runtime code execution.
No problem with applying your proposed change though.

Reply:
> I personally use the type parameter syntax to explicit the fact that the parameter is only used at `typescript` level and does not affect runtime code execution.

Since it is just a personal taste problem, let's stick w/ `<typeof rule>` approach then!

Another thought: this utility is really helpful, what about submitting a proposal to `@typescript-eslint` and adding this to `@typescript-eslint/rule-tester`?

Reply:
> this utility is really helpful, what about submitting a proposal to @typescript-eslint and adding this to @typescript-eslint/rule-tester?

Rather than adding an utility I think that a better approach would be to change the [`RuleTester#run`](https://github.com/typescript-eslint/typescript-eslint/blob/90617a651c3a8b244cdcc2a618aa999b7643ecce/packages/rule-tester/src/RuleTester.ts#L479-L486) method to infer options and messageIds from the `rule` parameter: something similar to `GetRuleType` used inside `createRuleTestCaseFunction`.
It would be a breaking change, but I think is better because it doesn't add the overhead of an additional utility function. 
Anyway I don't see a reason to not open an issue to propose this change or adding a new utility, 
thanks for the thought 😆!

I'll try to update the code here and submit the proposal to `@typescript-eslint/rule-tester` in the upcoming days.

Reply:
> It would be a breaking change, but I think is better because it doesn't add the overhead of an additional utility function.

We can do it gradually! In the meantime, we can export a `runTest` function instead of exporting a `runTester` object.

Reply:
I'm sorry, but I haven't understand what you mean: the test `utils/index` do not export any `RuleTester`. 
`RuleTester` instance are created each test file.
Also  `RuleTester#run` is not static method, it requires a `RuleTester` instance.

If you agree, let me update the code with the [extended InvalidTestCase](https://github.com/un-ts/eslint-plugin-import-x/pull/176#discussion_r1835906416)  and open the proposal to `@typescript-eslint`, then let's see what they say before doing additional changes here.

Reply:
Sure!

Reply:
Proposal created: https://github.com/typescript-eslint/typescript-eslint/issues/10321